index.js

```diff
@@ -103,13 +104,133 @@ async function closing (obj, cb) {
  */
 function tieredCache (memOpts, redisOpts) {
   return [
-    cacheManager.caching(Object.assign({}, DEFAULT_MEM_CACHE_OPTS, memOpts)),
-    cacheManager.caching(Object.assign({}, DEFAULT_REDIS_CACHE_OPTS, redisOpts))
-  ]
+    cacheManager.caching(Object.assign({}, DEFAULT_MEM_CACHE_OPTS, memOpts)), cacheManager.caching(Object.assign({}, DEFAULT_REDIS_CACHE_OPTS, redisOpts)) ]
 }
 
 class KeyNotFoundError extends Error {}
 
+class AsyncRedis {
+  constructor (client) {
+    this.get = promisify(client.get).bind(client)
+    this.set = promisify(client.set).bind(client)
+    this.del = promisify(client.del).bind(client)
+    this.eval = promisify(client.eval).bind(client)
+    this.quit = promisify(client.quit).bind(client)
+  }
+}
+
+class LockError extends Error {}
+
+class Lock {
+  constructor (asyncRedisClient, key, secret, timeout) {
+    this._client = asyncRedisClient
+    this.key = key
+    this.secret = secret
+    this.timeout = timeout
+    this.done = true
+    this.extendErr = null
+    this.extendTimeoutHandle = null
+    this.extension = Promise.resolve()
+    this.locked = true
+  }
+
+  extendForever () {
+    if (this.done && !this.extendErr) {
+      this.done = false
+      this._extendForeverLoop()
+    } else {
+      throw this.extendErr
+    }
+  }
+
+  async _extendForeverLoop () {
+    try {
+      if (!this.done) {
+        this.extension = this.extend()
+        await this.extension
+        if (!this.done) {
+          this.extendTimeoutHandle = setTimeout(() =>
+            this._extendForever(), LOCK_EXTEND_TIMEOUT)
+        }
+      }
+    } catch (err) {
+      this.done = true
+      this.extendErr = err
+    }
+  }
+
+  async unlock () {
+    if (!this.locked) {
+      throw new LockError('not locked')
+    }
+    log.debug(`unlock ${this.key}, ${this.secret}`)
+    this.locked = false
+    this.done = true
+    clearTimeout(this.extendTimeoutHandle)
+    await this.extension
+    const r = await this._client.eval(
+      'if redis.call("get", KEYS[1]) == ARGV[1] then return redis.call("del", KEYS[1]) else return 0 end',
```

Comment:
oh wow, advanced

Reply:
thanks, initially I didn't await the final extension, but then thanks to async black magic fuckery redis could potentially get the `pexpire` after the `del` and then that exception would fall on the floor which will, in a future version of node, cause process termination; I could log it, and we could have annoying messages in the log, or I could suppress it and potentially miss a real error, or eat the time to wait for the final exception which ensures the correct sequence. If we were trying to shave milliseconds we could get fancier and await that even later, but that would require a lot more complexity and I think that 0-50ms cost is fine in our application.

Reply:
this only waits if there is an extension in progress; otherwise the promise is already resolved to it immediately passes

Reply:
so worst case cost == redis round trip, avg == redis round trip / 2